test/Dialect/FIRRTL/grand-central.mlir

```diff
@@ -1123,6 +1123,220 @@ firrtl.circuit "SubmoduleDirectoryBehavior" attributes {
+  //   3) "C" indicates this is instantiated in the "Companion"
+  // E.g., "ET_C" is an external module instantiated above the DUT and in the
+  // Companion.
+  firrtl.module @MT__() {}
+  firrtl.module @M_D_() {}
+  firrtl.module @M__C() {}
+  firrtl.module @MTD_() {}
+  firrtl.module @M_DC() {}
+  firrtl.module @MT_C() {}
+  firrtl.module @MTDC() {}
+  firrtl.extmodule @ET__() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ET__.v", text = ""}
+  ]}
+  firrtl.extmodule @E_D_() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "E_D_.v", text = ""}
+  ]}
+  firrtl.extmodule @E__C() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "E__C.v", text = ""}
+  ]}
+  firrtl.extmodule @ETD_() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ETD_.v", text = ""}
+  ]}
+  firrtl.extmodule @E_DC() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "E_DC.v", text = ""}
+  ]}
+  firrtl.extmodule @ET_C() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ET_C.v", text = ""}
+  ]}
+  firrtl.extmodule @ETDC() attributes {annotations = [
+    {class = "firrtl.transforms.BlackBoxInlineAnno", name = "ETDC.v", text = ""}
+  ]}
+
+  // The Grand Central Companion module.
+  firrtl.module private @Companion() attributes {
+    annotations = [
+      {class = "sifive.enterprise.grandcentral.ViewAnnotation.companion",
+       defName = "Foo",
+       id = 0 : i64,
+       name = "View"}]} {
+
+    firrtl.instance m__c @M__C()
+    firrtl.instance m_dc @M_DC()
+    firrtl.instance mt_c @MT_C()
+    firrtl.instance mtdc @MTDC()
+
+    firrtl.instance e__c @E__C()
+    firrtl.instance e_dc @E_DC()
+    firrtl.instance et_c @ET_C()
+    firrtl.instance etdc @ETDC()
+  }
+
+  // The Design-under-test as indicated by the MarkDUTAnnotation
+  firrtl.module private @DUT() attributes {
+    annotations = [
+      {class = "sifive.enterprise.grandcentral.ViewAnnotation.parent",
+       id = 0 : i64,
+       name = "view"},
+      {class = "sifive.enterprise.firrtl.MarkDUTAnnotation"}
+    ]} {
+    firrtl.instance companion @Companion()
+
+    firrtl.instance m_d_ @M_D_()
+    firrtl.instance mtd_ @MTD_()
+    firrtl.instance m_dc @M_DC()
+    firrtl.instance mtdc @MTDC()
+
+    firrtl.instance e_d_ @E_D_()
+    firrtl.instance etd_ @ETD_()
+    firrtl.instance e_dc @E_DC()
+    firrtl.instance etdc @ETDC()
+  }
+
+  // The Top module that instantiates the DUT
+  firrtl.module @DirectoryBehaviorWithDUT() {
+    firrtl.instance dut @DUT()
+
+    firrtl.instance mt__ @MT__()
+    firrtl.instance mtd_ @MTD_()
+    firrtl.instance mt_c @MT_C()
+    firrtl.instance mtdc @MTDC()
+
+    firrtl.instance et__ @ET__()
+    firrtl.instance etd_ @ETD_()
+    firrtl.instance et_c @ET_C()
+    firrtl.instance etdc @ETDC()
+  }
+}
```

Comment:
This test is the following diagram. This is checking that "C" and "F" get directory information. These map to `*__C` and `*T_C`:

![image](https://user-images.githubusercontent.com/1018530/199359365-9dcd1c95-7a75-4fc0-9d3e-442d1553e7a8.png)